bigtable-dataflow-parent/bigtable-hbase-dataflow/src/main/java/com/google/cloud/bigtable/dataflow/CloudBigtableIO.java

```diff
@@ -748,11 +766,56 @@ void initializeScanner() throws IOException {
      */
     @Override
     public boolean advance() throws IOException {
-      current = scanIterator.next(scanner);
+      current = scanIterator.next(scanner, rangeTracker);
       rowsRead.addAndGet(scanIterator.getRowCount(current));
       return !scanIterator.isCompletionMarker(current);
     }
 
+
+    @Override
+    public final Double getFractionConsumed() {
+      return rangeTracker.getFractionConsumed();
+    }
+
+    /**
+     * Attempt to split the work by some percent of the ByteKeyRange based on a lexicographical
+     * split (and not statistics about the underlying table, which would be better, but that
+     * information does not exist).
+     */
+    @Override
+    public final synchronized BoundedSource<ResultOutputType> splitAtFraction(double fraction) {
+      ByteKey splitKey;
+      final ByteKeyRange range = source.getConfiguration().toByteKeyRange();
+      try {
+        splitKey = range.interpolateKey(fraction);
+      } catch (IllegalArgumentException e) {
+        READER_LOG.info("%s: Failed to interpolate key for fraction %s.", range, fraction);
+        return null;
+      }
+
+      READER_LOG.debug("Proposing to split {} at fraction {} (key {})", rangeTracker, fraction,
+        splitKey);
+
+      if (!rangeTracker.trySplitAtPosition(splitKey)) {
```

Comment:
should move this to the bottom, just before returning -- in case you get a transient exception creating sources or estimating size.